include/sound/sof/ext_manifest.h

```diff
@@ -99,4 +111,11 @@ struct ext_man_dbg_abi {
 	struct sof_ipc_user_abi_version dbg_abi;
 } __packed;
 
+/** EXT_MAN_ELEM_CONFIG_DATA elements (ABI3.17) */
+struct sof_ext_man_config_data {
+	struct sof_ext_man_elem_header hdr;
+
+	struct sof_config_elem elems[SOF_EXT_MAN_CONFIG_LAST_ELEM];
```

Comment:
same problem as in the `struct sof_ext_man_cavs_config_data` in your third patch? If I'm remembering correctly and we avoid variable length arrays because of a problem in xcc, then we probably should take @ranj063 's proposal to have some maximum element count here. Maybe
```
#define SOF_CONFIG_MAX_NUMBER ALIGN(SOF_EXT_MAN_CONFIG_LAST_ELEM, 8)
```
or something similar? In general - these arrays aren't required to contain elements of all types, right? Some might be missing?

Reply:
yes, xcc's has problem with variable size arrays (eg. https://github.com/thesofproject/sof/pull/2522) - it's why there is defined ELEM_LAST (in FW enum) used to define array capacity. Empty elements (with 0-value) should be just skipped, we don't need to worry about memory consumption here - it's never placed in runtime DSP memory. It's why first token's enum value starts from 1. Introducing `SOF_CONFIG_MAX_NUMBER` significant educe extension possibility of this dictionary.
In conclusion, I think it's good as it is, and empty elements should be just skipped.

Reply:
> yes, xcc's has problem with variable size arrays (eg. [thesofproject/sof#2522](https://github.com/thesofproject/sof/pull/2522)) - it's why there is defined ELEM_LAST (in FW enum) used to define array capacity. Empty elements (with 0-value) should be just skipped, we don't need to worry about memory consumption here - it's never placed in runtime DSP memory. It's why first token's enum value starts from 1.

@ktrzcinx well, it's up to us how to identify "empty" elements - whether by a 0 in field `.a` or by `-EINVAL` in field `.b`

> Introducing `SOF_CONFIG_MAX_NUMBER` significant reduce extension possibility of this dictionary.
> In conclusion, I think it's good as it is, and empty elements should be just skipped.

Sorry, I don't see how `SOF_CONFIG_MAX_NUMBER` reduces extensibility. It reduces the *need* to extend the array by giving you a few reserved elements. E.g. now with only one element defined, you'd allocate an array of 8 (or 4). So, you don't need to extend the array for the next 7 elements. If we ever come round to defining the ninth element, then it's a change in the ABI, we add 8 more elements. So, you have to modify the structure only once out of 8 times.

Reply:
> yes, xcc's has problem with variable size arrays (

A reminder that variable size arrays and flexible array members are two VERY different things. Better use the correct name when referring to compiler bugs.

Variable size arrays look like this and their memory is managed automagically:
```C
func(uint8_t n)
{
   var_size_array[n];
   ...   
}
```

Flexible array members look like this, _their memory must be managed by the programmer_:
```C
struct something {
  ...
  flex_array_must_be_last[];
};

```

One of them is now forbidden in the Linux kernel, the other is not.

Reply:
@lyakh @ktrzcinx I updated with defined MAX_NUMBER. This is pairing with https://github.com/thesofproject/sof/pull/3510

Reply:
@fredoh9 I also have some solution for the size , see https://github.com/thesofproject/sof/pull/3527 this will reduce our change to minimum.. 
Agreed with @kv2019i in https://github.com/thesofproject/linux/pull/2459 @ktrzcinx @fredoh9 I think you may need to decided who's PR goes first. We'd better make some alignment with header first.

Reply:
@fredoh9 sorry, I don't see the update that you're talking about? I still see `LAST_ELEM` which I still don't find a particularly good solution.